File: packages/data-ops/src/contracts/research.ts
import { z } from "zod";

import { projectCategorySchema } from "./desktop-api/project";
import { referoContextSchema } from "./refero";

/** Codex/engine emits explicit null for absent fields — use nullish, not optional-only. */
const optionalText = z.string().min(1).nullish();
const optionalUrl = z.string().min(1).nullish();

export const DETAILS_STRUCTURE_SECTIONS = [
  "CTA",
  "Footer",
  "Hero",
  "Legal",
  "Navigation",
  "Drawer",
  "Dropdown",
  "Fullscreen",
  "Morphing",
] as const;

export const DETAILS_PAGE_SECTIONS = [
  "404",
  "Article",
  "Blog",
  "Case Study",
  "Contact",
  "Content",
  "About",
  "FAQ",
  "Features",
  "Services",
  "Steps",
  "Newsletter",
  "Portfolio",
  "Pricing",
  "Products",
  "Social Proof",
  "Logo",
  "Testimonial",
  "Stats",
  "Team",
  "Timeline",
] as const;

export const DETAILS_SECTIONS = [
  ...DETAILS_STRUCTURE_SECTIONS,
  ...DETAILS_PAGE_SECTIONS,
] as const;
export const REFERO_APP_SECTIONS = [
  "Onboarding",
  "Login / Sign up",
  "Homepage",
  "Dashboard",
  "Analytics / Reports",
  "Table / List",
  "Search",
  "Detail View",
  "Forms",
  "Settings",
  "Profile",
  "Team / Permissions",
  "Integrations",
  "Billing",
  "Pricing",
  "Checkout",
  "Browse / Discovery",
  "Notifications",
  "Empty State",
  "Success / Confirmation",
  "Splash Screen",
] as const;
export const REFERO_WEB_APP_SECTIONS = [
  "Onboarding",
  "Login / Sign up",
  "Homepage",
  "Dashboard",
  "Analytics / Reports",
  "Table / List",
  "Search",
  "Detail View",
  "Forms",
  "Settings",
  "Profile",
  "Team / Permissions",
  "Integrations",
  "Billing",
  "Pricing",
  "Checkout",
  "Notifications",
  "Empty State",
  "Success / Confirmation",
] as const;
export const REFERO_IOS_APP_SECTIONS = [
  "Splash Screen",
  "Onboarding",
  "Login / Sign up",
  "Homepage",
  "Browse / Discovery",
  "Search",
  "Detail View",
  "Dashboard",
  "Analytics / Reports",
  "Forms",
  "Pricing",
  "Checkout",
  "Profile",
  "Settings",
  "Notifications",
  "Empty State",
  "Success / Confirmation",
] as const;
export const RESEARCH_REFERENCE_SECTIONS = [
  ...DETAILS_SECTIONS,
  "Onboarding",
  "Login / Sign up",
  "Homepage",
  "Checkout",
  "Dashboard",
  "Analytics / Reports",
  "Table / List",
  "Search",
  "Detail View",
  "Forms",
  "Settings",
  "Profile",
  "Team / Permissions",
  "Integrations",
  "Billing",
  "Browse / Discovery",
  "Notifications",
  "Empty State",
  "Success / Confirmation",
  "Splash Screen",
] as const;
export const DEFAULT_DETAILS_SECTIONS = [
  "Hero",
  "Features",
  "Social Proof",
  "Pricing",
  "Contact",
] as const;
export const DEFAULT_REFERO_APP_SECTIONS = [
  "Onboarding",
  "Homepage",
  "Pricing",
  "Checkout",
  "Dashboard",
] as const;
export const detailsSectionSchema = z.enum(RESEARCH_REFERENCE_SECTIONS);

export const researchMatrixScoreSchema = z.enum(["Strong", "OK", "Weak"]);

export const researchArtifactSectionSchema = z.enum([
  "summary",
  "companySnapshot",
  "competitiveAnalysis",
  "uiPatterns",
  "targetUsers",
  "opportunities",
]);

export const researchInputSchema = z.object({
  projectId: z.string().min(1),
  projectName: z.string().min(1),
  projectCategory: projectCategorySchema,
  clientName: z.string().min(1).optional(),
  industry: z.string().min(1),
  website: z.string().min(1).optional(),
  projectBrief: z.string().min(1).optional(),
  competitorUrls: z.array(z.string().min(1)).default([]),
  detailsSections: z.array(detailsSectionSchema).min(1).default([...DEFAULT_DETAILS_SECTIONS]),
  targetUsers: z.string().min(1).optional(),
  additionalNotes: z.string().min(1).optional(),
  uploadedAssetIds: z.array(z.string().min(1)).default([]),
});

export const researchCompanySnapshotRowSchema = z.object({
  label: z.string().min(1),
  value: z.string().min(1),
});

export const researchCompetitorSchema = z.object({
  id: z.string().min(1),
  name: z.string().min(1),
  url: optionalText,
  logoUrl: optionalUrl,
  mark: optionalText,
  color: optionalText,
  positioning: optionalText,
  summary: optionalText,
  strengths: z.array(z.string().min(1)).default([]),
  weaknesses: z.array(z.string().min(1)).default([]),
  sourceReferenceIds: z.array(z.string().min(1)).default([]),
});

export const researchCompetitiveMatrixCellSchema = z.object({
  competitorId: z.string().min(1),
  score: researchMatrixScoreSchema,
  note: z.string().nullish(),
});

export const researchCompetitiveMatrixRowSchema = z.object({
  id: z.string().min(1),
  label: z.string().min(1),
  cells: z.array(researchCompetitiveMatrixCellSchema).default([]),
});

export const researchCompetitiveAnalysisSchema = z.object({
  competitors: z.array(researchCompetitorSchema).default([]),
  matrixRows: z.array(researchCompetitiveMatrixRowSchema).default([]),
});

export const researchUiPatternExampleSchema = z.object({
  id: z.string().min(1),
  title: z.string().min(1),
  imageUrl: optionalUrl,
  thumbnailUrl: optionalUrl,
  sourceProduct: optionalText,
  sourceReferenceId: optionalText,
});

export const researchUiPatternGroupSchema = z.object({
  id: z.string().min(1),
  title: z.string().min(1),
  summary: optionalText,
  patternCountLabel: optionalText,
  recognizedPatterns: z.array(z.string().min(1)).default([]),
  examples: z.array(researchUiPatternExampleSchema).default([]),
});

export const researchTargetUserSchema = z.object({
  id: z.string().min(1),
  name: z.string().min(1),
  role: z.string().min(1),
  goals: z.array(z.string().min(1)).default([]),
  frustrations: z.array(z.string().min(1)).default([]),
  context: optionalText,
  relevance: optionalText,
  assumptions: z.array(z.string().min(1)).default([]),
});

export const researchOpportunitySchema = z.object({
  id: z.string().min(1),
  title: optionalText,
  description: z.string().min(1),
  sourceSection: researchArtifactSectionSchema.nullish(),
});

export const researchCustomSectionSchema = z.object({
  id: z.string().min(1),
  title: z.string().min(1),
  body: z.string().min(1),
});

export const researchSourceReferenceSchema = z.object({
  id: z.string().min(1),
  provider: z.enum(["refero", "details", "figma", "notion", "sheets", "website", "user"]),
  label: z.string().min(1),
  url: optionalUrl,
  externalId: optionalText,
});

export const researchArtifactSchema = z.object({
  apiVersion: z.literal("v1"),
  artifactKind: z.literal("researchArtifact"),
  projectId: z.string().min(1),
  title: z.string().min(1),
  summary: z.array(z.string().min(1)).min(1).default([]),
  companySnapshot: z.array(researchCompanySnapshotRowSchema).min(1).default([]),
  competitiveAnalysis: researchCompetitiveAnalysisSchema,
  uiPatterns: z.array(researchUiPatternGroupSchema).default([]),
  uiPatternProvider: z.enum(["refero", "details"]).optional(),
  targetUsers: z.array(researchTargetUserSchema).min(1).default([]),
  opportunities: z.array(researchOpportunitySchema).min(1).default([]),
  customSections: z.array(researchCustomSectionSchema).default([]),
  openQuestions: z.array(z.string().min(1)).default([]),
  sourceReferences: z.array(researchSourceReferenceSchema).default([]),
  referoContext: referoContextSchema.optional(),
  generatedAt: z.number().int().nonnegative(),
});

export const researchArtifactPatchSchema = z.object({
  artifactId: z.string().min(1),
  section: researchArtifactSectionSchema,
  patch: z.unknown(),
  reason: z.string().min(1).optional(),
});

export type ResearchMatrixScore = z.infer<typeof researchMatrixScoreSchema>;
export type DetailsSection = z.infer<typeof detailsSectionSchema>;
export type ResearchArtifactSection = z.infer<typeof researchArtifactSectionSchema>;
export type ResearchInput = z.infer<typeof researchInputSchema>;
export type ResearchCompanySnapshotRow = z.infer<typeof researchCompanySnapshotRowSchema>;
export type ResearchCompetitor = z.infer<typeof researchCompetitorSchema>;
export type ResearchCompetitiveMatrixCell = z.infer<typeof researchCompetitiveMatrixCellSchema>;
export type ResearchCompetitiveMatrixRow = z.infer<typeof researchCompetitiveMatrixRowSchema>;
export type ResearchCompetitiveAnalysis = z.infer<typeof researchCompetitiveAnalysisSchema>;
export type ResearchUiPatternExample = z.infer<typeof researchUiPatternExampleSchema>;
export type ResearchUiPatternGroup = z.infer<typeof researchUiPatternGroupSchema>;
export type ResearchTargetUser = z.infer<typeof researchTargetUserSchema>;
export type ResearchOpportunity = z.infer<typeof researchOpportunitySchema>;
export type ResearchCustomSection = z.infer<typeof researchCustomSectionSchema>;
export type ResearchSourceReference = z.infer<typeof researchSourceReferenceSchema>;
export type ResearchArtifact = z.infer<typeof researchArtifactSchema>;
export type ResearchArtifactPatch = z.infer<typeof researchArtifactPatchSchema>;

File: apps/stage-engine/src/models/research.rs
#![allow(dead_code)]
// Contract mirror for Research artifacts. The endpoint wiring lands after the
// Refero and Claude/Codex boundaries are validated.

use serde::{Deserialize, Serialize};

use super::refero::ReferoContext;

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub enum ProjectCategory {
    #[serde(rename = "websites", alias = "web-design")]
    Websites,
    #[serde(rename = "web-apps", alias = "web-app")]
    WebApps,
    #[serde(rename = "ios-apps", alias = "app-design")]
    IosApps,
}

impl ProjectCategory {
    pub fn display_name(self) -> &'static str {
        match self {
            Self::Websites => "Websites",
            Self::WebApps => "Web apps",
            Self::IosApps => "iOS apps",
        }
    }
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub enum ResearchMatrixScore {
    Strong,
    OK,
    Weak,
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum ResearchArtifactSection {
    Summary,
    CompanySnapshot,
    CompetitiveAnalysis,
    UiPatterns,
    TargetUsers,
    Opportunities,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchInput {
    pub project_id: String,
    pub project_name: String,
    pub client_name: Option<String>,
    pub project_category: ProjectCategory,
    pub industry: String,
    pub website: Option<String>,
    pub project_brief: Option<String>,
    #[serde(default)]
    pub competitor_urls: Vec<String>,
    #[serde(default)]
    pub details_sections: Vec<String>,
    pub target_users: Option<String>,
    pub additional_notes: Option<String>,
    #[serde(default)]
    pub uploaded_asset_ids: Vec<String>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchCompanySnapshotRow {
    pub label: String,
    pub value: String,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchCompetitor {
    pub id: String,
    pub name: String,
    pub url: Option<String>,
    pub logo_url: Option<String>,
    pub mark: Option<String>,
    pub color: Option<String>,
    pub positioning: Option<String>,
    pub summary: Option<String>,
    #[serde(default)]
    pub strengths: Vec<String>,
    #[serde(default)]
    pub weaknesses: Vec<String>,
    #[serde(default)]
    pub source_reference_ids: Vec<String>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchCompetitiveMatrixCell {
    pub competitor_id: String,
    pub score: ResearchMatrixScore,
    pub note: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchCompetitiveMatrixRow {
    pub id: String,
    pub label: String,
    #[serde(default)]
    pub cells: Vec<ResearchCompetitiveMatrixCell>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchCompetitiveAnalysis {
    #[serde(default)]
    pub competitors: Vec<ResearchCompetitor>,
    #[serde(default)]
    pub matrix_rows: Vec<ResearchCompetitiveMatrixRow>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchUiPatternExample {
    pub id: String,
    pub title: String,
    pub image_url: Option<String>,
    pub source_product: Option<String>,
    pub source_reference_id: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchUiPatternGroup {
    pub id: String,
    pub title: String,
    pub summary: Option<String>,
    pub pattern_count_label: Option<String>,
    #[serde(default)]
    pub recognized_patterns: Vec<String>,
    #[serde(default)]
    pub examples: Vec<ResearchUiPatternExample>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchTargetUser {
    pub id: String,
    pub name: String,
    pub role: String,
    #[serde(default)]
    pub goals: Vec<String>,
    #[serde(default)]
    pub frustrations: Vec<String>,
    pub context: Option<String>,
    pub relevance: Option<String>,
    #[serde(default)]
    pub assumptions: Vec<String>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchOpportunity {
    pub id: String,
    pub title: Option<String>,
    pub description: String,
    pub source_section: Option<ResearchArtifactSection>,
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum ResearchSourceProvider {
    Refero,
    Figma,
    Notion,
    Sheets,
    Website,
    Details,
    User,
}

#[derive(Clone, Copy, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum ResearchUiPatternProvider {
    #[default]
    Refero,
    Details,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchSourceReference {
    pub id: String,
    pub provider: ResearchSourceProvider,
    pub label: String,
    pub url: Option<String>,
    pub external_id: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchArtifact {
    pub api_version: String,
    pub artifact_kind: String,
    pub project_id: String,
    pub title: String,
    #[serde(default)]
    pub summary: Vec<String>,
    #[serde(default)]
    pub company_snapshot: Vec<ResearchCompanySnapshotRow>,
    pub competitive_analysis: ResearchCompetitiveAnalysis,
    #[serde(default)]
    pub ui_patterns: Vec<ResearchUiPatternGroup>,
    #[serde(default)]
    pub ui_pattern_provider: ResearchUiPatternProvider,
    #[serde(default)]
    pub target_users: Vec<ResearchTargetUser>,
    #[serde(default)]
    pub opportunities: Vec<ResearchOpportunity>,
    #[serde(default)]
    pub open_questions: Vec<String>,
    #[serde(default)]
    pub source_references: Vec<ResearchSourceReference>,
    pub refero_context: Option<ReferoContext>,
    pub generated_at: u128,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchArtifactPatch {
    pub artifact_id: String,
    pub section: ResearchArtifactSection,
    pub patch: serde_json::Value,
    pub reason: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::ProjectCategory;

    #[test]
    fn project_category_serializes_with_canonical_ids() {
        assert_eq!(
            serde_json::to_string(&ProjectCategory::Websites).expect("serialize category"),
            "\"websites\""
        );
        assert_eq!(
            serde_json::to_string(&ProjectCategory::WebApps).expect("serialize category"),
            "\"web-apps\""
        );
        assert_eq!(
            serde_json::to_string(&ProjectCategory::IosApps).expect("serialize category"),
            "\"ios-apps\""
        );
    }

    #[test]
    fn project_category_accepts_legacy_aliases() {
        assert_eq!(
            serde_json::from_str::<ProjectCategory>("\"web-design\"")
                .expect("deserialize legacy website category"),
            ProjectCategory::Websites
        );
        assert_eq!(
            serde_json::from_str::<ProjectCategory>("\"web-app\"")
                .expect("deserialize legacy web app category"),
            ProjectCategory::WebApps
        );
        assert_eq!(
            serde_json::from_str::<ProjectCategory>("\"app-design\"")
                .expect("deserialize legacy iOS category"),
            ProjectCategory::IosApps
        );
    }

    #[test]
    fn project_category_rejects_unknown_values() {
        assert!(serde_json::from_str::<ProjectCategory>("\"branding\"").is_err());
    }
}
